Update README.md to include the GHES migration path (#249)
github/gh-gei: README.md

* Azure DevOps -> GitHub Enterprise Cloud migrations
* GitHub Enterprise Cloud -> GitHub Enterprise Cloud migrations

> * GitHub Enterprise Server -> GitHub Enterprise Cloud migrations

Author: Is it too soon to call this GHES->GHEC a "supported" path?

Reviewer: might be worth adding (>3.4)

to ensure it is noted for customers.

> * GitHub Enterprise Server (> version 3.4) -> GitHub Enterprise Cloud migrations

Reviewer 2: ```suggestion
* GitHub Enterprise Server (version 3.4 or newer) -> GitHub Enterprise Cloud migrations
```

Reviewer 2: I'd make this more clear as some customers get confused with `>` and `<` in the past experience.

  -?, -h, --help  Show help and usage information

Commands:
  generate-script       Generates a migration script. This provides you the ability to review the steps that this tool will take, and optionally modify the script if desired before running it.
  grant-migrator-role   Allows an organization admin to grant a USER or TEAM the migrator role for a single GitHub organization. The migrator role allows the role assignee to perform migrations into the target organization.
                        Note: Expects GH_PAT env variable to be set.
  migrate-repo          Invokes the GitHub APIs to migrate the repo and all repo data.
  revoke-migrator-role  Allows an organization admin to revoke the migrator role for a USER or TEAM for a single GitHub organization. This will remove their ability to run a migration into the target organization.
                        Note: Expects GH_PAT env variable to be set.
```

To generate a migration script, you'll need to set `GH_PAT` as an environment variable for your destination and `GH_SOURCE_PAT` for your source location. 

## GitHub Enterprise Server (GHES) to GitHub Migration Usage

> GEI-CLI is a cross-platform .NET Core console application. General usage will use the `generate-script` command to create a script that can be used to migrate all repositories from a GHES organization.

Reviewer 3: Looks like this line is a dupe from above

Author: That line is also at the start of the ADO section. I assumed it was there in case someone looking to do an ADO migration isn't reading the gh->gh section, so I added it for GHES too

Reviewer 3: Hmm, I think the first sentence is unnecessary at repeating, but the second sentence could make sense to have. Does that sound reasonable to you ?

Author: Threw it up to the top, lemme know what you think https://github.com/github/gh-gei/pull/249/commits/632c1326cad86d6d8a77e1abba54d9e5ad95cc2c

Due to many GHES instances containing firewall rules restricting the direct API access that GEI needs to perform a migration, the GEI CLI uploads the migration's archive data to Azure Blob Storage as a step prior to starting the migration using GEI. This allows GEI to perform a migration without directly accessing the GHES instance. Because of this extra step, you'll pass in a connection string to an Azure Storage account, specified in the parameters below.

### Command Line

Generating a migration script for GHES is very similar to GitHub to GitHub migration scripts, but requires a couple extra parameters. You can run `gh gei generate-script --help` to learn about all of the options.

The extra options you will need when migrating from GHES are:

> - `--ghes-api-url`

Reviewer 3: Maybe worth providing an example?

Author: I didn't want to copy+paste the same description that the help output gives, but I could throw an example of the whole command below these?

Reviewer 3: I like that idea!

Author: Added in https://github.com/github/gh-gei/pull/249/commits/a1a5add3736f4e1282fd7a022ead154ae1031c2b

> - `--azure-storage-connection-string`

Reviewer 3: For this it's probably worth calling out you'll have to use single (or double) quotes around the value because connection string itself uses characters that would otherwise be interpreted by bash

Author: Updated in https://github.com/github/gh-gei/pull/249/commits/0024e771b4b0e70240142b282e46267edeac56e6

  grant-migrator-role   Allows an organization admin to grant a USER or TEAM the migrator role for a single GitHub organization. The migrator role allows the role assignee to perform migrations into the target organization.
                        Note: Expects GH_PAT env variable to be set.
  migrate-repo          Invokes the GitHub APIs to migrate the repo and all repo data.
  revoke-migrator-role  Allows an organization admin to revoke the migrator role for a USER or TEAM for a single GitHub organization. This will remove their ability to run a migration into the target organization.
                        Note: Expects GH_PAT env variable to be set.
```

To generate a migration script, you'll need to set `GH_PAT` as an environment variable for your destination and `GH_SOURCE_PAT` for your source location. 

## GitHub Enterprise Server (GHES) to GitHub Migration Usage

General usage will use the `generate-script` command to create a script that can be used to migrate all repositories from a GHES organization.

> Due to many GHES instances containing firewall rules restricting the direct API access that GEI needs to perform a migration, the GEI CLI uploads the migration's archive data to Azure Blob Storage as a step prior to starting the migration using GEI. This allows GEI to perform a migration without directly accessing the GHES instance. Because of this extra step, you'll pass in a connection string to an Azure Storage account, specified in the parameters below.

Reviewer 2: Should add a few pointers: 

* Reference to creating a blob storage container: https://docs.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-portal
* How to get the connection string: https://docs.microsoft.com/en-us/azure/data-explorer/kusto/api/connection-strings/storage-connection-strings

Author: Updated in https://github.com/github/gh-gei/pull/249/commits/54d28940de6d2dcc19b43775774e76fa07441cbc

Author: Updated in https://github.com/github/gh-gei/pull/249/commits/54d28940de6d2dcc19b43775774e76fa07441cbc

> Due to many GHES instances containing firewall rules restricting the direct API access that GEI needs to perform a migration, the GEI CLI uploads the migration's archive data to Azure Blob Storage as a step prior to starting the migration using GEI. This allows GEI to perform a migration without directly accessing the GHES instance. Because of this extra step, you'll pass in a connection string to an Azure Storage account, specified in the parameters below.

Reviewer 2: ```suggestion
Due to many GHES instances containing firewall rules restricting the direct API access that GEI needs to perform a migration, the GEI CLI uploads the migration's archive data to Azure Blob Storage as a step prior to starting the migration using GEI. This allows GEI to perform a migration without directly accessing the GHES instance. To get started you'll need to setup an [Azure Blob storage container](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-portal). Then you'll pass in a [connection string](https://docs.microsoft.com/en-us/azure/data-explorer/kusto/api/connection-strings/storage-connection-strings) to that Azure Storage account, specified in the parameters below.
```

Author: Updated in https://github.com/github/gh-gei/pull/249/commits/54d28940de6d2dcc19b43775774e76fa07441cbc

### Command Line

Generating a migration script for GHES is very similar to GitHub to GitHub migration scripts, but requires a couple extra parameters. You can run `gh gei generate-script --help` to learn about all of the options.

The extra options you will need when migrating from GHES are:

> - `--ghes-api-url`

Reviewer 2: We should include the text from the CLI here.

Author: Updated in https://github.com/github/gh-gei/pull/249/commits/636443ee5314cee758bb71ace3763a45e541d3fe

- `--azure-storage-connection-string` - it's recommend to use quotes around this value since the connection string itself uses characters that might otherwise be interpreted by bash

> - `--no-ssl-verify` (optional)

Reviewer 2: We should include the text from the CLI here.

Author: I thought copy+pasting from the CLI would be redundant and lead to discrepancies if we made tweaks to the descriptions without updating the readme, but I can do that 👍

Here's an example:
```
gh gei generate-script --github-source-org "source-ghes-org" --github-target-org "target-ghec-org" --ghes-api-url "https://api.myghesinstance.com" --azure-storage-connection-string "DefaultEndpointsProtocol=https;AccountName=myazureaccount;AccountKey=myazurestoragekey;EndpointSuffix=core.windows.net" --no-ssl-verify
```

> This will generate a new file (named `./migrate.ps1` by default) which gives you an opportunity to review the steps GEI-CLI will take. Once ready, running the `./migrate.ps1` file (in PowerShell, for example) will kick off the migration!

Reviewer 2: ```suggestion
This will generate a new file (named `./migrate.ps1` by default) which gives you an opportunity to review the steps GEI-CLI will take. Once ready, running the `./migrate.ps1` file will kick off the migration!
```
